[single-shot] Added open/close for the new single-shot implementation

kparichay opened the pull request:
Added open/close for the new single-shot implementation
New single-shot directly uses tensor-filter subplugin to avoid gstreamer overhead
Get input/output info is also added
The API for the new implementation is not changed.

Working on #1547

Self evaluation:
Build test: [x]Passed [ ]Failed [ ]Skipped
Run test: [x]Passed [ ]Failed [ ]Skipped

Signed-off-by: Parichay Kapoor pk.kapoor@samsung.com

jaeyun-jung commented on api/capi/src/nnstreamer-capi-single-new.c:
```diff
@@ -0,0 +1,500 @@
+      } else if (g_str_has_suffix (path_down, ".pb")) {
+        ml_logi ("The given model [%s] is supposed a tensorflow model.", model);
+        *nnfw = ML_NNFW_TYPE_TENSORFLOW;
+      } else if (!g_str_has_suffix (path_down, ".so")) {
```
line 99 and 109
let's consider to use NNSTREAMER_SO_FILE_EXTENSION for so-file extension.

kparichay replied:
Updated.

jaeyun-jung commented on api/capi/src/nnstreamer-capi-single-new.c:
```diff
@@ -0,0 +1,500 @@
+  if (status != ML_ERROR_NONE)
+    return status;
+
+  if (!g_file_test (model, G_FILE_TEST_IS_REGULAR)) {
```
I found that this can be moved to line 86, before checking the file extension. (added a PR to update this for old single-shot function.)
Plz consider to move this block.

kparichay replied:
Yes, you are right
Updated.

jaeyun-jung commented on api/capi/src/nnstreamer-capi-single-new.c:
```diff
@@ -0,0 +1,500 @@
+    status = ML_ERROR_INVALID_PARAMETER;
+    goto error;
+  }
+  status = klass->start (single_h->filter);
```
klass->start() does not return ml-error code. IMO it just checks bool result like line 282. 
!klass->start (single_h->filter)

kparichay replied:
Updated to directly check the bool result.

kparichay commented:
V2:
Replace `.so` with `NNSTREAMER_SO_FILE_EXTENSION`
Corrected location of model check in `ml_single_check_nnfw`
Corrected result check for `klass->start`

jaeyun-jung commented on api/capi/src/nnstreamer-capi-single-new.c:
```diff
@@ -0,0 +1,499 @@
+  if (!g_file_test (model, G_FILE_TEST_IS_REGULAR)) {
+    ml_loge ("The given param, model path [%s] is invalid.",
+        GST_STR_NULL (model));
+    status = ML_ERROR_INVALID_PARAMETER;
```
just return ML_ERROR_INVALID_PARAMETER

kparichay replied:
updated.

jaeyun-jung commented on api/capi/src/nnstreamer-capi-single-new.c:
```diff
@@ -0,0 +1,500 @@
+  }
+
+  g_free (path_down);
+  if (status != ML_ERROR_NONE)
```
also this condition is unnecessary

kparichay replied:
updated.

jaeyun-jung commented on api/capi/src/nnstreamer-capi-single-new.c:
```diff
@@ -0,0 +1,498 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
```
change license (Apache -> LGPL-2.1)

kparichay replied:
updated.

jaeyun-jung commented on api/capi/meson.build:
```diff
@@ -94,6 +81,56 @@ nnstreamer_capi_dep = declare_dependency(link_with: nnstreamer_capi_lib,
 )
 
+# Create dependency for tensor_filter_single for single-shot capi
+tensor_filter_single_lib = shared_library ('tensor_filter_single_lib',
```
Question about tensor_filter_single.c :
Is it necessary to build shared library for _single.c? How about adding this file to capi_single_new_main?

kparichay replied:
adding `tensor_filter_single.c` to `capi_single_new_main` would expose `g_tensor_filter_single_g_type` with `libcapi-nnstreamer-single-new.so`

myungjoo replied:
You are exposing them anyway as long as the application writer has access to C/C++ shared libraries.
As long as the capi-dev provides a header that lists public-APIs only, it's fine; SDK will block functions calls to be not exposed, anyway.

So.. please provide a single .so

kparichay replied:
done.

jaeyun-jung commented on api/capi/src/nnstreamer-capi-single-new.c:
```diff
@@ -0,0 +1,496 @@
+    if (ml_single_get_input_info (single_h, &in_info) != ML_ERROR_NONE)
+      goto error;
+    if (ml_tensors_info_clone (&single_h->in_info, in_info) != ML_ERROR_NONE)
+      goto error;
```
need destroy()
ml_tensors_info_destroy (in_info);

kparichay replied:
updated.

jaeyun-jung commented on api/capi/src/nnstreamer-capi-single-new.c:
```diff
@@ -0,0 +1,496 @@
+    if (ml_single_get_output_info (single_h, &out_info) != ML_ERROR_NONE)
+      goto error;
+    if (ml_tensors_info_clone (&single_h->out_info, out_info) != ML_ERROR_NONE)
+      goto error;
```
same here 
ml_tensors_info_destroy (out_info);

kparichay replied:
updated.

leemgs commented on api/capi/src/nnstreamer-capi-single-new.c:
```diff
@@ -0,0 +1,504 @@
+}
+
+/**
+ * @brief Opens an ML model and returns the instance as a handle.
```
Looks good.

kparichay commented:
V3:
`libcapi-nnstreamer-single-new.so` is the only .so provided
`libtensor_filter_single.so` has been merged with it

myungjoo commented on packaging/nnstreamer.spec:
```diff
@@ -327,6 +341,18 @@ cp -r result %{buildroot}%{_datadir}/nnstreamer/unittest/
+%files -n capi-nnstreamer-single-new-devel
+%{_includedir}/nnstreamer/nnstreamer.h
+%{_includedir}/nnstreamer/nnstreamer-single.h
+%{_libdir}/pkgconfig/capi-nnstreamer.pc
```
You can't test this library with this .pc file. (it won't direct libcapi-nnstreamer-single-new.so)

kparichay replied:
updated.

dongju-chae commented on api/capi/meson.build:
```diff
@@ -94,6 +81,43 @@ nnstreamer_capi_dep = declare_dependency(link_with: nnstreamer_capi_lib,
+  dependencies: capi_deps,
+  include_directories: [
+    inc,
+    nninc
```
how about declaring a variable for include_directories (because it's repeated)? like `capi_inc`

kparichay replied:
done

kparichay commented on api/capi/src/nnstreamer-capi-single-new.c:
```diff
@@ -0,0 +1,504 @@
+  /* Set input option */
+  str_dim = gst_tensors_info_get_dimensions_string (&info);
+  str_type = gst_tensors_info_get_types_string (&info);
+  str_name = gst_tensors_info_get_names_string (&info);
```
Good catch, updated with the check

dongju-chae replied:
it has still problems. g_free (NULL) is gonna happen.

kparichay replied:
https://developer.gnome.org/glib/stable/glib-Memory-Allocation.html#g-free
`g_free` can handle NULL by itself

dongju-chae replied:
Oh, g_free ignores NULL. never mind.
